clash-lib/src/Clash/Normalize/Types.hs

```diff
@@ -28,25 +29,25 @@ import Clash.Rewrite.Types    (Rewrite, RewriteMonad)
 -- | State of the 'NormalizeMonad'
 data NormalizeState
   = NormalizeState
-  { _normalized          :: BindingMap
+  { _normalized          :: MVar BindingMap
```

Comment:
This is insufficient because [`inlineWorkFree`](https://github.com/clash-lang/clash-compiler/blob/f39fb9805e27745177939eb269427ac482648a53/clash-lib/src/Clash/Normalize/Transformations/Inline.hs#L637-L711) [calls](https://github.com/clash-lang/clash-compiler/blob/f39fb9805e27745177939eb269427ac482648a53/clash-lib/src/Clash/Normalize/Transformations/Inline.hs#L705) [`normalizeTopLvlBndr`](https://github.com/clash-lang/clash-compiler/blob/f39fb9805e27745177939eb269427ac482648a53/clash-lib/src/Clash/Normalize/Util.hs#L419-L458)

So why is this bad? Take the following example:
```
top x = (g x, f x)

f x = h + x
{-# NOINLINE f #-}
g x = h * x
{-# NOINLINE g #-}

h = expensive_workfree_expression
```
After normalizing `top`, we normalize `f` and `g` concurrently. And both will do an `inlineWorkFree` of `h`. They could then both start normalizing `h` before committing it to the cache. Thus I think `normalizeTopLvlBndr` should do something akin:

```
normalizeTopLvlBndr isTop nm (Binding nm' sp inl pr tm _) = do
  normalizedV <- Lens.use (extra.normalized)
  cache <- takeMVar normalizedV
  case lookup nm cache of
    Just vMVar ->
       putMVar normalizedV cache
       readMVar vMVar
    Nothing ->
      tmp <- newEmptyMVar
      putMVar normalizedV (extendVarEnv nm tmp cache)
      ... continue normalization ...
      putMVar tmp value
      return value
```
I doesn't save us much time, because the thread arriving second still has to wait for the first. But the further along the first thread is, the more time we save.

Reply:
Done! That is clever.

It slows things down a little bit, however, perhaps due to more `MVar`s or perhaps because I convert to lists in order to `mapM` over a `VarEnv`.